File: gui/baculum/protected/Portlets/JobList.php
<?php
/*
 * Bacula(R) - The Network Backup Solution
 * Baculum   - Bacula web interface
 *
 * Copyright (C) 2013-2015 Marcin Haba
 *
 * The main author of Baculum is Marcin Haba.
 * The original author of Bacula is Kern Sibbald, with contributions
 * from many others, a complete list can be found in the file AUTHORS.
 *
 * You may use this file and others of this release according to the
 * license defined in the LICENSE file, which includes the Affero General
 * Public License, v3.0 ("AGPLv3") and some additional permissions and
 * terms pursuant to its AGPLv3 Section 7.
 *
 * This notice must be preserved when any source code is
 * conveyed and/or propagated.
 *
 * Bacula(R) is a registered trademark of Kern Sibbald.
 */

Prado::using('System.Web.UI.ActiveControls.TActiveDataGrid');
Prado::using('System.Web.UI.ActiveControls.TActiveRepeater');
Prado::using('System.Web.UI.ActiveControls.TActivePanel');
Prado::using('System.Web.UI.ActiveControls.TCallback');
Prado::using('Application.Portlets.ISlideWindow');
Prado::using('Application.Portlets.Portlets');

class JobList extends Portlets implements ISlideWindow {

	public $ID;
	public $buttonID;
	public $windowTitle;
	public $jobLevels;
	public $jobStates;
	public $runningJobStates;
	public $jobTypes;

	public function setID($id) {
		$this->ID = $id;
	}

	public function getID($hideAutoID = true) {
		return $this->ID;
	}

	public function setButtonID($id) {
		$this->buttonID = $id;
	}

	public function getButtonID() {
		return $this->buttonID;
	}

	public function setWindowTitle($param) {
		$this->windowTitle = $param;
	}

	public function getWindowTitle() {
		return $this->windowTitle;
	}

	public function onLoad($param) {
		parent::onLoad($param);
		$allowedButtons = array('JobBtn', 'ReloadJobs', 'Run', 'RunJobAgain');
		if($this->Page->IsPostBack || $this->Page->IsCallBack) {
			if(in_array($this->getPage()->CallBackEventTarget->ID, $allowedButtons)) {
				$this->prepareData();
			}
		}
		$misc = $this->Application->getModule('misc');
		$this->jobLevels = $misc->getJobLevels();
		$this->jobStates = $misc->getJobState();
		$this->jobTypes = $misc->getJobType();
		$this->runningJobStates = $misc->getRunningJobStates();
	}

	public function prepareData($forceReload = false) {
		$allowedButtons = array('JobBtn', 'ReloadJobs', 'Run', 'RunJobAgain');
		if($this->Page->IsPostBack || $this->Page->IsCallBack || $forceReload) {
			if(in_array($this->getPage()->CallBackEventTarget->ID, $allowedButtons) || $forceReload) {
				$params = $this->getUrlParams('jobs', $this->getPage()->JobWindow->ID);
				$jobs = $this->Application->getModule('api')->get($params);
				$isDetailView = $_SESSION['view' . $this->getPage()->JobWindow->ID] == 'details';
				if($isDetailView === true) {
					$this->RepeaterShow->Visible = false;
					$this->DataGridShow->Visible = true;
					$this->DataGrid->DataSource = $this->Application->getModule('misc')->objectToArray($jobs->output);
					$this->DataGrid->dataBind();
				} else {
					$this->RepeaterShow->Visible = true;
					$this->DataGridShow->Visible = false;
					$this->Repeater->DataSource = $jobs->output;
					$this->Repeater->dataBind();
				}
			}
		}
	}

	public function sortDataGrid($sender, $param) {
		$params = $this->getUrlParams('jobs', $this->getPage()->JobWindow->ID);
		$data = $this->Application->getModule('api')->get($params)->output;
		$data = $this->Application->getModule('misc')->objectToArray($data);
		$this->DataGrid->DataSource = $this->sortData($data, $param->SortExpression, $sender->UniqueID);
		$this->DataGrid->dataBind();
	}

	public function configure($sender, $param) {
		if($this->Page->IsCallBack) {
			$this->getPage()->JobConfiguration->configure($param->CallbackParameter);
		}
	}

	public function run_again($sender, $param) {
		if($this->Page->IsCallBack) {
			$this->getPage()->JobConfiguration->run_again(null, $param->CallbackParameter);
		}
	}

	public function executeAction($action) {
		$params = explode(';', $this->CheckedValues->Value);
		$commands = array();
		switch($action) {
			case 'delete': {
				for($i = 0; $i < count($params); $i++) {
					$cmd = array('delete');
					$cmd[] = 'jobid="' . $params[$i] . '"';
					$cmd[] = 'yes';
					$cmd[] = PHP_EOL;
					$commands[] = implode(' ', $cmd);
				}
				$this->getPage()->Console->CommandLine->Text = implode(' ', $commands);
				$this->getPage()->Console->sendCommand(null, null);
				break;
			}
		}
		$this->CheckedValues->Value = "";
	}

	public function formatJobName($name) {
		if (strlen($name) > 24) {
			$name = substr($name, 0, 10) . '...' . substr($name, -11);
		}
		return $name;
	}
}
?>

File: bacula/src/tools/joblist.c
/*
   Bacula(R) - The Network Backup Solution

   Copyright (C) 2000-2020 Kern Sibbald

   The original author of Bacula is Kern Sibbald, with contributions
   from many others, a complete list can be found in the file AUTHORS.

   You may use this file and others of this release according to the
   license defined in the LICENSE file, which includes the Affero General
   Public License, v3.0 ("AGPLv3") and some additional permissions and
   terms pursuant to its AGPLv3 Section 7.

   This notice must be preserved when any source code is
   conveyed and/or propagated.

   Bacula(R) is a registered trademark of Kern Sibbald.
*/

#include "bacula.h"
#include "lib/unittests.h"
#include "filed/fd_plugins.h"
#include "findlib/bfile.h"

/* Pointers to Bacula functions */
static const  char *working="./";

#define USE_JOB_LIST
#define UNITTESTS
#include "plugins/fd/fd_common.h"

void *start_heap;

int main (int argc, char *argv[])
{
   Unittests t("joblist_test");
   POOL_MEM tmp;
   const char *k = "netapp:/vol/vol1";
   const char *k2 = "netapp:/vol/vol2";
   const char *k3 = "netapp:/vol/vol2xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx";
   const char *k4 = "netapp:/vol/vol2xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxz";
   char key1[128], key2[128], key3[128], key4[128], key5[128],
      jobF[128], jobI[128], jobI2[128], jobD[128], jobI3[128];
   strcpy(key1, "ID=704ba403-fa5d-45f9-8681-859eec950723 DATE=1:15f18c5c-3824-40d2-8843-e6d161f77504:15f18c5c-3824-40d2-8843-e6d161f77504");
   strcpy(key2, "ID=704ba403-fa5d-45f9-8681-859eec950723 DATE=2:15f18c5c-3824-40d2-8843-e6d161f77504:15f18c5c-3824-40d2-8843-e6d161f77504");
   strcpy(key3, "ID=704ba403-fa5d-45f9-8681-859eec950723 DATE=3:15f18c5c-3824-40d2-8843-e6d161f77504:15f18c5c-3824-40d2-8843-e6d161f77504");
   strcpy(key4, "ID=704ba403-fa5d-45f9-8681-859eec950723 DATE=4:15f18c5c-3824-40d2-8843-e6d161f77504:15f18c5c-3824-40d2-8843-e6d161f77504");
   strcpy(key5, "ID=704ba403-fa5d-45f9-8681-859eec950723 DATE=5:15f18c5c-3824-40d2-8843-e6d161f77504:15f18c5c-3824-40d2-8843-e6d161f77504");

   strcpy(jobF, "Test1.2020-03-12_09.45.31_24");
   strcpy(jobI, "Test1.2020-03-12_09.45.31_25");
   strcpy(jobI2, "Test1.2020-03-12_09.45.31_26");
   strcpy(jobD, "Test1.2020-03-12_09.45.31_27");
   strcpy(jobI3, "Test1.2020-03-12_09.45.31_28");

   if (argc != 1) {
      exit (1);
   }
   unlink("test.dat");
   {
      joblist job_history(NULL, k, jobF, NULL, 'F');
      job_history.set_base("test.dat");
      job_history.store_job(key1);
      job_history.prune_jobs(NULL, NULL, NULL);
   }
   {
      joblist job_history(NULL, k, jobI, jobF, 'I');
      job_history.set_base("test.dat");
      ok(job_history.find_job(jobF, tmp.handle()), "Find previous job");
      ok(strcmp(tmp.c_str(), key1) == 0, "test key");
      job_history.prune_jobs(NULL, NULL, NULL);
      job_history.store_job(key2);
   }
   {
      joblist job_history(NULL, k, jobI2, jobI, 'I');
      job_history.set_base("test.dat");
      ok(job_history.find_job(jobF, tmp.handle()), "Find previous job");
      ok(strcmp(tmp.c_str(), key1) == 0, "test key");
      ok(job_history.find_job(jobI, tmp.handle()), "Find previous job");
      ok(strcmp(tmp.c_str(), key2) == 0, "test key");
      job_history.prune_jobs(NULL, NULL, NULL);
      job_history.store_job(key3);
   }
   {
      joblist job_history(NULL, k, jobD, jobF, 'D');
      job_history.set_base("test.dat");
      ok(job_history.find_job(jobF, tmp.handle()), "Find previous job");
      ok(strcmp(tmp.c_str(), key1) == 0, "test key");
      job_history.prune_jobs(NULL, NULL, NULL);
      job_history.store_job(key4);
   }
   {
      joblist job_history(NULL, k, jobI3, jobD, 'I');
      job_history.set_base("test.dat");
      ok(job_history.find_job(jobF, tmp.handle()), "Find previous job");
      ok(strcmp(tmp.c_str(), key1) == 0, "test key");
      job_history.prune_jobs(NULL, NULL, NULL);
      job_history.store_job(key5);
      nok(job_history.find_job(jobI, tmp.handle()), "Find pruned job");
   }

   {
      joblist job_history(NULL, k2, jobF, NULL, 'F');
      job_history.set_base("test.dat");
      job_history.store_job(key1);
      job_history.prune_jobs(NULL, NULL, NULL);
   }
   {
      joblist job_history(NULL, k2, jobI, jobF, 'I');
      job_history.set_base("test.dat");
      ok(job_history.find_job(jobF, tmp.handle()), "Find previous job");
      ok(strcmp(tmp.c_str(), key1) == 0, "test key");
      job_history.prune_jobs(NULL, NULL, NULL);
      job_history.store_job(key2);
   }
   {
      joblist job_history(NULL, k2, jobI2, jobI, 'I');
      job_history.set_base("test.dat");
      ok(job_history.find_job(jobF, tmp.handle()), "Find previous job");
      ok(strcmp(tmp.c_str(), key1) == 0, "test key");
      ok(job_history.find_job(jobI, tmp.handle()), "Find previous job");
      ok(strcmp(tmp.c_str(), key2) == 0, "test key");
      job_history.prune_jobs(NULL, NULL, NULL);
      job_history.store_job(key3);
   }
   {
      joblist job_history(NULL, k2, jobD, jobF, 'D');
      job_history.set_base("test.dat");
      ok(job_history.find_job(jobF, tmp.handle()), "Find previous job");
      ok(strcmp(tmp.c_str(), key1) == 0, "test key");
      job_history.prune_jobs(NULL, NULL, NULL);
      job_history.store_job(key4);
   }
   {
      joblist job_history(NULL, k2, jobI3, jobD, 'I');
      job_history.set_base("test.dat");
      ok(job_history.find_job(jobF, tmp.handle()), "Find previous job");
      ok(strcmp(tmp.c_str(), key1) == 0, "test key");
      job_history.prune_jobs(NULL, NULL, NULL);
      job_history.store_job(key5);
      nok(job_history.find_job(jobI, tmp.handle()), "Find pruned job");
   }
   {
      joblist job_history(NULL, k3, jobF, NULL, 'F');
      job_history.set_base("test.dat");
      job_history.store_job(key1);
      job_history.prune_jobs(NULL, NULL, NULL);
   }
   {
      joblist job_history(NULL, k3, jobI, jobF, 'I');
      job_history.set_base("test.dat");
      ok(job_history.find_job(jobF, tmp.handle()), "Find previous very long job");
      ok(strcmp(tmp.c_str(), key1) == 0, "test key");
      job_history.prune_jobs(NULL, NULL, NULL);
      job_history.store_job(key2);
   }
   {
      joblist job_history(NULL, k4, jobI, jobF, 'I');
      job_history.set_base("test.dat");
      nok(job_history.find_job(jobF, tmp.handle()), "Find previous very long job");
   }
   return report();
}
